helix-core/src/test/java/org/apache/helix/manager/zk/TestRawZkClient.java

```diff
@@ -461,4 +467,247 @@ void testPendingRequestGauge() throws Exception {
+        1, // resources
+        10, // partitions per resource
+        5, // number of nodes
+        3, // replicas
+        "MasterSlave",
+        true); // do rebalance
+
+    final long originalSessionId = _zkClient.getSessionId();
+    final String originalHexSessionId = _zkClient.getHexSessionId();
+    final String path = "/" + methodName;
+
+    // Verify the node is not existed.
+    Assert.assertFalse(_zkClient.exists(path));
+
+    // Expire the original session.
+    ZkTestHelper.expireSession(_zkClient);
+
+    // Wait until the ZkClient has got a new session.
+    Assert.assertTrue(TestHelper.verify(() -> {
+      try {
+        // New session id should be greater than expired session id.
+        return _zkClient.getSessionId() > originalSessionId;
+      } catch (HelixException ex) {
+        return false;
+      }
+    }, 1000L));
+
+    try {
+      // Try to create ephemeral node with the original session.
+      // This creation should NOT be successful because the original session is already expired.
+      _zkClient.createEphemeral(path, "Hello Helix", originalHexSessionId);
+      Assert.fail("Ephemeral node should not be created by the expired session.");
+    } catch (ZkSessionMismatchedException expected) {
+      // Expected because there is a session mismatch.
+    }
+
+    // Verify the node is not created.
+    Assert.assertFalse(_zkClient.exists(path));
+  }
+
+  /*
+   * Tests that when trying to create an ephemeral node, if connection to zk service is lost,
+   * ConnectionLossException will be thrown in retryUntilConnected and ephemeral creation will fail.
+   * Because ConnectionLossException is not thrown in createEphemeral(), operation retry timeout is
+   * set to 3 seconds and then ZkTimeoutException is thrown. And retry cause message is checked to
+   * see if ConnectionLossException was thrown before retry.
+   */
+  @Test(timeOut = 5 * 60 * 1000L)
+  public void testConnectionLossWhileCreateEphemeral() throws Exception {
+    final String methodName = TestHelper.getTestMethodName();
+
+    final ZkClient zkClient = new ZkClient.Builder()
+        .setZkServer(ZK_ADDR)
+        .setOperationRetryTimeout(3000L) // 3 seconds
+        .build();
+
+    final String expectedSessionId = zkClient.getHexSessionId();
+    final String path = "/" + methodName;
+    final String data = "data";
+
+    Assert.assertFalse(zkClient.exists(path));
+
+    // Shutdown zk server so zk operations will fail due to disconnection.
+    TestHelper.stopZkServer(_zkServer);
+
+    try {
+      final CountDownLatch countDownLatch = new CountDownLatch(1);
+      final AtomicBoolean running = new AtomicBoolean(true);
+
+      final Thread creationThread = new Thread(() -> {
+        while (running.get()) {
+          try {
+            // Create ephemeral node in the expected session id.
+            zkClient.createEphemeral(path, data, expectedSessionId);
+          } catch (ZkTimeoutException e) {
```

Comment:
the connectionloss keeper exception is not concerted to ZKTimeoutException. But each time before retrying, the operation time is checked. If timed out, ZKTimeoutException is thrown. And this is why I add the retry cause message to the exception: to help this test, as well.
```
        // before attempting a retry, check whether retry timeout has elapsed
        if (System.currentTimeMillis() - operationStartTime > _operationRetryTimeoutInMillis) {
          throw new ZkTimeoutException("Operation cannot be retried because of retry timeout ("
              + _operationRetryTimeoutInMillis + " milli seconds). Retry was caused by "
              + retryCauseCode);
        }
```